### PR TITLE
perf(extmarks): allow ephemeral extmarks past EOF

### LINKED ISSUE #14756: tree-sitter incredible slow with single line big json due to hotspot in nvim_buf_set_extmark
- `nvim --version`:   
v0.5.0-dev+1357-g192f89ea1
                               Build type: RelWithDebInfo
                                                         LuaJIT 2.1.0-beta3

- Operating system/version:
archlinux x64
- Terminal name/version:
xterm(also happened when using nvim-qt)

[ ] `vim -u DEFAULTS` (version: ) behaves differently (not tried)

### Steps to reproduce using `nvim -u NORC`
only happen if using with nvim-treesitter, while using nvim -u NORC, nvim highlight head parts of json and not hang.

### Actual behaviour

while using with tree-sitter, nvim incredible slow when opening very big json file(in one line),  profile shows nvim_buf_set_extmark consume most of CPU time. after a while, even after file is highlighted, cursor move still very slow, especially when cursor hit object start '{'

### Expected behaviour

tree-sitter highlight should be responsive.

minimal config:
```
set nocompatible
filetype off


call plug#begin('~/.vim/bundle')

"vim-scripts

Plug 'nvim-treesitter/nvim-treesitter', {'do': ':TSUpdate'}
call plug#end()


lua << EOF


require'nvim-treesitter.configs'.setup {
    ensure_installed = "maintained",
    highlight = {
        enable = true
    },
    indent = {
        enable = true
    }
}

EOF

```

while nvim hanging, I've run profiler attached, profile result shows that nvim_buf_set_extmark consume 96% cpu time.
source code shows that src/nvim/api/buffer.c 1497 and 1687 each consume about 50% cpu time.
```
  1492   size_t len = 0;
  1493   if (line < 0 || line > buf->b_ml.ml_line_count) {
  1494     api_set_error(err, kErrorTypeValidation, "line value outside range");
  1495     return 0;
  1496   } else if (line < buf->b_ml.ml_line_count) {
  1497     len = STRLEN(ml_get_buf(buf, (linenr_T)line+1, false));             <------------------------ hotspot1
  1498   }



  1685  if (col2 >= 0) {
  1686     if (line2 >= 0 && line2 < buf->b_ml.ml_line_count) {
  1687       len = STRLEN(ml_get_buf(buf, (linenr_T)line2 + 1, false));          <--------------------------hotspot2
  1688     } else if (line2 == buf->b_ml.ml_line_count) {
  1689       // We are trying to add an extmark past final newline
  1690       len = 0;
  1691     } else {
  1692       // reuse len from before
  1693       line2 = (int)line;
  1694     }


```

### PR DESCRIPTION
**Pull request by vigoux**
Should partially fix #14756

cc @bfredl

### CODE REVIEW

File: src/nvim/api/buffer.c [hunk 1]
```diff
@@ -1683,7 +1683,7 @@ Integer nvim_buf_set_extmark(Buffer buffer, Integer ns_id,
   }
 
   if (col2 >= 0) {
-    if (line2 >= 0 && line2 < buf->b_ml.ml_line_count) {
+    if (!ephemeral && line2 >= 0 && line2 < buf->b_ml.ml_line_count) {
```
Comment (bfredl) [inline src/nvim/api/buffer.c:1686]: IMHO a separate `if (ephemeral) { len = MAXCOL; } ` branch is cleaner

**Comment by bfredl** [comment]
you also need to disable the `STRLEN` call for `col` argument earlier.

**Comment by vigoux** [comment]
Yeah but, I am not really sure how to do it, either I move all the argument checking below the option checking (may be time consuming then), or I just set it to a default value, what would be better?

**Comment by bfredl** [comment]
just move the column check with STRLEN below option parsing. it only makes a difference for the error path where perf doesn't matter.

**Comment by vigoux** [comment]
This new version should fix #14756 completely then.
@comicfans could you confirm it is the case ?

File: src/nvim/api/buffer.c [hunk 2]
```diff
@@ -1674,6 +1659,22 @@ Integer nvim_buf_set_extmark(Buffer buffer, Integer ns_id,
     }
   }
 
+  size_t len = 0;
+  if (line < 0 || line > buf->b_ml.ml_line_count) {
+    api_set_error(err, kErrorTypeValidation, "line value outside range");
+    return 0;
+  } else if (!ephemeral && line < buf->b_ml.ml_line_count) {
```
Comment (bfredl) [inline src/nvim/api/buffer.c:1666]: will this actually work? maybe 

```
} else if (line < buf->b_ml.ml_line_count) {
  len = ephemeral ? MAXCOL : STRLEN(ml_get_buf(buf, (linenr_T)line+1, false));
}
```

Comment (vigoux) [reply src/nvim/api/buffer.c:1666]: Seems better indeed.

Comment (vigoux) [reply src/nvim/api/buffer.c]: Updated

File: src/nvim/api/buffer.c (hunk 1 above)
Comment (bfredl) [reply src/nvim/api/buffer.c:1687]: maybe could repeat the same pattern here? :)

Comment (vigoux) [reply src/nvim/api/buffer.c:1687]: Yup

**Comment by comicfans** [comment]
Hi @vigoux, I've tested this branch, now nvim can open big json immediately (also highlighted), and original hotspot already disappeared from profiler, so I think original problem (slow highlight) is resolved. but I realized moving cursor still very slow, and nvim still hang if using '%' to jump to corresponding json mark, new hotspot in profiler, I'll create a new issue for it

**Comment by vigoux** [comment]
Great to see it's fixed for the highlight part, for the cursor, we'll see in the other issue (feel free to ping me in it) 
@bfredl is it okay for you, I think it should be added for 0.5 as it's a performance fix.

**Pull request merged**